### PR TITLE
Automatically install openshift internal registry certs in serving operator

### PR DESCRIPTION
**Pull request by savitaashture**
After installing Knative Serving it should allow to pull images from the OpenShift internal registry using serving operator.

Jira:
https://jira.coreos.com/browse/SRVKS-115
https://jira.coreos.com/browse/SRVKS-103

### CODE REVIEW

File: deploy/resources/knative-serving-0.6.0.yaml [hunk 1]
```diff
@@ -1656,3 +1656,10 @@ spec:
         name: config-logging
 
 ---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  labels:
+    serving.knative.dev/release: "v0.6.0"
+  name: config-service-ca
+  namespace: knative-serving
```
Comment (jcrossley3) [inline deploy/resources/knative-serving-0.6.0.yaml]: I would like to keep this file as close to identical to upstream's 0.6.0 release as possible, especially when it comes to platform-specific features (openshift). You're already creating CM in the operator, so there should be no need for it to be here, right?

File: pkg/controller/install/add_openshift.go [hunk 2]
```diff
@@ -5,7 +5,7 @@ import (
 )
 
 func init() {
-	platformPreInstallFuncs = append(platformPreInstallFuncs, openshift.EnsureMaistra)
-	platformPostInstallFuncs = append(platformPostInstallFuncs, openshift.EnsureOpenshiftIngress)
+	platformPreInstallFuncs = append(platformPreInstallFuncs, openshift.EnsureMaistra, openshift.EnsureConfigMapForCrt)
+	platformPostInstallFuncs = append(platformPostInstallFuncs, openshift.EnsureOpenshiftIngress, openshift.EnsureOpenshiftRegistry)
```
Comment (jcrossley3) [inline pkg/controller/install/add_openshift.go]: I hate to do this to you, but I've completely changed this in master. Hopefully, it's simpler, but you'll need to rebase this PR.

File: pkg/controller/install/openshift/openshift.go [hunk 3]
```diff
@@ -86,6 +87,168 @@ func EnsureOpenshiftIngress(c client.Client, scheme *runtime.Scheme, instance *s
...
 
+// EnsureConfigMapForCrt ensures crt injection into the configmap
+func EnsureConfigMapForCrt(c client.Client, scheme *runtime.Scheme, instance *servingv1alpha1.Install) error {
+	namespace := instance.GetNamespace()
+	if routeExists, err := kindExists(c, "route", "route.openshift.io/v1", namespace); err != nil {
+		return err
+	} else if !routeExists {
+		// Not running in OpenShift
+		return nil
+	}
```
Comment (jcrossley3) [inline pkg/controller/install/openshift/openshift.go]: Checking for the route shouldn't be necessary anymore, which renders this function superfluous. You just need the `updateConfigMapForCrt` now.

File: pkg/controller/install/openshift/openshift.go [hunk 4]
```diff
@@ -86,6 +87,168 @@ func EnsureOpenshiftIngress(c client.Client, scheme *runtime.Scheme, instance *s
...
+
+	if err := updateConfigMapForCrt(c, namespace); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func updateConfigMapForCrt(c client.Client, namespace string) error {
```
Comment (jcrossley3) [inline pkg/controller/install/openshift/openshift.go]: Once you rebase, you'll see that this "preInstall" function's signature should just expect an Install instance, from which you can obtain the namespace. And there's a package var called `api` you can use for your client calls.

File: pkg/controller/install/openshift/openshift.go [hunk 5]
```diff
@@ -86,6 +87,168 @@ func EnsureOpenshiftIngress(c client.Client, scheme *runtime.Scheme, instance *s
...
+			if err != nil {
+				return err
+			}
+			// ConfigMap created successfully
+			return nil
+		}
+		return err
+	}
+
+	/* Check existence of service-ca.crt in the configMap
```
Comment (jcrossley3) [inline pkg/controller/install/openshift/openshift.go]: Why is this necessary?

File: pkg/controller/install/openshift/openshift.go [hunk 6]
```diff
@@ -86,6 +87,168 @@ func EnsureOpenshiftIngress(c client.Client, scheme *runtime.Scheme, instance *s
...
+		if err := c.Update(context.TODO(), cm); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// EnsureOpenshiftRegistry ensures registry for openshift
+func EnsureOpenshiftRegistry(c client.Client, scheme *runtime.Scheme, instance *servingv1alpha1.Install) error {
```
Comment (jcrossley3) [inline pkg/controller/install/openshift/openshift.go]: Same comment regarding the function signature, and all these functions can be private now, btw. Only `Configure` needs to be public.

File: pkg/controller/install/openshift/openshift.go [hunk 7]
```diff
@@ -86,6 +87,168 @@ func EnsureOpenshiftIngress(c client.Client, scheme *runtime.Scheme, instance *s
...
+	namespace := instance.GetNamespace()
+	if routeExists, err := kindExists(c, "route", "route.openshift.io/v1", namespace); err != nil {
+		return err
+	} else if !routeExists {
+		// Not running in OpenShift
+		return nil
+	}
```
Comment (jcrossley3) [inline pkg/controller/install/openshift/openshift.go]: No longer necessary to check for a route. `Configure` handles that.

File: pkg/controller/install/openshift/openshift.go [hunk 8]
```diff
@@ -86,6 +87,168 @@ func EnsureOpenshiftIngress(c client.Client, scheme *runtime.Scheme, instance *s
...
+
+	if err := updateConfigMapForRegistry(c, namespace); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func updateDeploymentController(c client.Client, namespace string) error {
```
Comment (jcrossley3) [inline pkg/controller/install/openshift/openshift.go]: Now that we have `Configure`, you don't even need `EnsureOpenshiftRegistry` any more. Just this one, with a slightly different signature.

File: pkg/controller/install/openshift/openshift.go [hunk 9]
```diff
@@ -86,6 +87,168 @@ func EnsureOpenshiftIngress(c client.Client, scheme *runtime.Scheme, instance *s
...
+
+		if err := c.Update(context.TODO(), found); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func updateConfigMapForRegistry(c client.Client, namespace string) error {
```
Comment (jcrossley3) [inline pkg/controller/install/openshift/openshift.go]: This should be a `Transformer` like `ingress` or `egress`. Because this CM is already a part of the upstream manifest, we can just change its values _before_ we apply it. It should greatly simplify the code as well.

File: deploy/resources/knative-serving-0.6.0.yaml (hunk 1 above)
Comment (savitaashture) [reply deploy/resources/knative-serving-0.6.0.yaml]: Iam keeping this here because in transform function for all the files ownerReference will be added So if i create a configmap in install function then ownerReference "serving.knative.dev/v1alpha1" will not be added for newly created configmap...
So in the operator iam just updating configmap with annotation and checking the condition whether configmap exist or not

File: pkg/controller/install/add_openshift.go (hunk 2 above)
Comment (savitaashture) [reply pkg/controller/install/add_openshift.go]: Okey i will rebase and change accordingly...

File: pkg/controller/install/openshift/openshift.go (hunk 3 above)
Comment (savitaashture) [reply pkg/controller/install/openshift/openshift.go]: Okey...

File: pkg/controller/install/openshift/openshift.go (hunk 4 above)
Comment (savitaashture) [reply pkg/controller/install/openshift/openshift.go]: Yes..I have changed according to the latest code.

File: pkg/controller/install/openshift/openshift.go (hunk 5 above)
Comment (savitaashture) [reply pkg/controller/install/openshift/openshift.go]: I have changed logic so this condition no more exist.

I  added that condition to check if at all  CM exist then that condition should check existence of key and update the annotation.

File: pkg/controller/install/openshift/openshift.go (hunk 6 above)
Comment (savitaashture) [reply pkg/controller/install/openshift/openshift.go]: Handled...

File: pkg/controller/install/openshift/openshift.go (hunk 7 above)
Comment (savitaashture) [reply pkg/controller/install/openshift/openshift.go]: Okey

File: pkg/controller/install/openshift/openshift.go (hunk 8 above)
Comment (savitaashture) [reply pkg/controller/install/openshift/openshift.go]: Modified according to the latest code base.

File: pkg/controller/install/openshift/openshift.go (hunk 9 above)
Comment (savitaashture) [reply pkg/controller/install/openshift/openshift.go]: Handled..Thank you.

File: deploy/resources/knative-serving-0.6.0.yaml (hunk 1 above)
Comment (jcrossley3) [reply deploy/resources/knative-serving-0.6.0.yaml]: I understand your logic, but we don't want the CM in this file. :)

This file should match the upstream release artifact. That's why `0.6.0` is in its name. The only time we might change something in here is if we need to cherry-pick some upstream fix prior to the next release.

Remember that this operator might be used to install KS on any number of different platforms (GKE, Azure, minikube, etc) but only OpenShift is going to do anything with a CM named `config-service-ca`, and that Azure user is going to wonder why it's there. :)

So I'm afraid you'll need to change `ensureConfigMapForCrt` back from a Transformer to a PreInstall hook. And you can set that owner reference as easy as you're setting that annotation:

    import "k8s.io/apimachinery/pkg/apis/meta/v1"
    func ensureConfigMapForCrt(instance *servingv1alpha1.Install) error {
        ...
        cm.SetOwnerReferences([]v1.OwnerReference{*v1.NewControllerRef(instance, instance.GroupVersionKind())})
        ...

Make sense?

Comment (savitaashture) [reply deploy/resources/knative-serving-0.6.0.yaml]: Okey thank you i understood the context i will change and update the code..

**Comment by jcrossley3** [comment]
Looks great! Thanks!

**Comment by savitaashture** [comment]
@jcrossley3 Thank you

File: pkg/controller/install/openshift/openshift.go [hunk 10]
```diff
@@ -178,6 +240,39 @@ func egress(u *unstructured.Unstructured) *unstructured.Unstructured {
 	return u
 }
 
+func ensureOpenShiftRegistryForConfigMap(u *unstructured.Unstructured) *unstructured.Unstructured {
+	if u.GetKind() == "ConfigMap" && u.GetName() == "config-deployment" {
+		data := map[string]string{"registriesSkippingTagResolving": "ko.local,dev.local,docker-registry.default.svc:5000,image-registry.openshift-image-registry.svc:5000"}
```
Comment (bbrowning) [inline pkg/controller/install/openshift/openshift.go]: We should not be editing the list of tag resolving registries here. Doing so actually makes the SSL related changes irrelevant. Instead, we want the SSL changes in this PR but should not touch this config setting at all with our operator.
